README.md
# perprof-web

<!--
[![Tests](https://github.com/abelsiqueira/perprof-web/actions/workflows/tests.yml/badge.svg)](https://github.com/abelsiqueira/perprof-web/actions/workflows/tests.yml)
[![Lint](https://github.com/abelsiqueira/perprof-web/actions/workflows/lint.yml/badge.svg)](https://github.com/abelsiqueira/perprof-web/actions/workflows/lint.yml)
![PyPI](https://img.shields.io/pypi/v/perprof-web)
[![](https://img.shields.io/badge/docs-latest-blue)](https://abelsiqueira.github.io/perprof-web/latest)
[![](https://img.shields.io/badge/docs-dev-blue)](https://abelsiqueira.github.io/perprof-web/dev)
-->

Web interface for the [perprof-py](https://github.com/abelsiqueira/perprof-py) for creating performance profiles (as described by [Dolan and Moré](http://arxiv.org/abs/cs/0102001)).

## Reference

When using this software for publications, please cite the paper below, which
describes this project:

> Siqueira, A. S., Costa da Silva, R. G. and Santos, L.-R., (2016). Perprof-web: A Python Package for Performance Profile of Mathematical Optimization Software. Journal of Open Research Software. 4(1), p.e12. DOI: [10.5334/jors.81](http://doi.org/10.5334/jors.81).

## Install

Use `perprof-web` from Docker if you just want to use it.

## How to use

Run perprof-web using Docker with

```bash
docker run -p 5000:5000 abelsiqueira/perprof-web
```

## Getting Help

If you didn't find something at the documentation, want to report a bug, or request a new feature, please open an [issue](https://github.com/abelsiqueira/perprof-web/issues).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.
Internal changes will frequently not be included here.

The format is based on [Keep a Changelog],
and this project adheres to [Semantic Versioning].

## [unreleased]

## [0.0.2]

### Added

- Create Docker image

## [0.0.1]

### Added

- Initial version

<!-- Links -->
[keep a changelog]: https://keepachangelog.com/en/1.0.0/
[semantic versioning]: https://semver.org/spec/v2.0.0.html

<!-- Versions -->
<!-- markdown-link-check-disable -->
[unreleased]: https://github.com/abelsiqueira/perprof-web/compare/v0.0.2...HEAD
[0.0.2]: https://github.com/abelsiqueira/perprof-web/releases/tag/v0.0.2
[0.0.1]: https://github.com/abelsiqueira/perprof-web/releases/tag/v0.0.1


perprof_web/views.py
"""Main file."""


import os

from flask import redirect, render_template, request, url_for
from perprof.solver_data import SolverData, read_table
from werkzeug.utils import secure_filename

from perprof_web import app
from perprof_web.plotting import matplotlib_performance_profile

solvers = []
# For debugging purposes:
# solvers = [
#     read_table(os.path.join(app.config["UPLOAD_FOLDER"], "alpha.table")),
#     read_table(os.path.join(app.config["UPLOAD_FOLDER"], "beta.table")),
#     read_table(os.path.join(app.config["UPLOAD_FOLDER"], "gamma.table")),
# ]

plot_options = {
    "title": "Performance Profile",
    "xlabel": "Performance ratio",
    "ylabel": "Percentage of problems solved",
    "semilogx": "checked",
    "width": "800",
    "height": "500",
}


@app.route("/")
def home():
    """Home."""
    return render_template(
        "index.html",
        solvers=solvers,
    )


@app.route("/solvers", methods=("POST", "GET"))
def show_solvers():
    """Main page for solvers."""
    if request.method == "POST":
        action = request.form.get("action")

        if action == "upload":
            # upload file flask
            uploaded_file = request.files["uploaded-file"]
            filename = secure_filename(uploaded_file.filename)
            savepath = os.path.join(app.config["UPLOAD_FOLDER"], filename)
            uploaded_file.save(savepath)

            ext = filename.split(".")[-1]
            if ext == "csv":
                solvers.append(
                    SolverData(
                        f"Unnamed solver {len(solvers) + 1}",
                        savepath,
                    )
                )
            elif ext == "table":
                solvers.append(read_table(savepath))
            else:
                # TODO: Treat correctly using Flask
                raise ValueError(f"Bad {filename}. Unexpected extension {ext}.")
        elif action == "update":
            for field in ["title", "xlabel", "ylabel", "width", "height"]:
                current_value = plot_options[field]
                new_value = request.form.get(field)
                if new_value != current_value:
                    plot_options[field] = new_value

            plot_options["semilogx"] = request.form.get("semilogx")
        else:
            raise ValueError(f"Unexpected action {action}")

    plot_data = matplotlib_performance_profile(solvers, plot_options)

    return render_template(
        "solvers.html",
        plot_data=plot_data,
        plot_options=plot_options,
        solvers=solvers,
    )


@app.route("/solvers/<int:solver_id>")
def show_solver(solver_id):
    """Details of a solver."""
    return render_template(
        "solver.html",
        solvers=solvers,
        solver=solvers[solver_id],
    )


@app.route("/update/solver/<int:solver_id>", methods=["POST"])
def update_solver(solver_id):
    """Delete or update a solver."""
    action = request.form.get("action")

    if action == "delete":
        solvers.pop(solver_id)
    elif action == "edit":
        new_algname = request.form.get("algname")
        if len(new_algname.strip()) == 0:
            return "Solver name must be non-null"
        if new_algname != solvers[solver_id].algname:
            solvers[solver_id].algname = new_algname

    return redirect(url_for("show_solvers"))


perprof_web/plotting.py
"""Functions related to plotting."""

import base64
from io import BytesIO

from matplotlib.figure import Figure
from perprof.profile_data import ProfileData

DPI = 96


def matplotlib_performance_profile(solvers, plot_options):
    """Plot the performance profile using matplotlib.

    Args:
        solvers (list[SolverData]): List of solvers
        plot_options (dict):
            Options, I guess. The following options are expected:

            - 'title'
            - 'xlabel'
            - 'ylabel'
            - 'semilogx' (Boolean): Is it a semilogx plot?

    Returns:
        plot_data: Data converted into a base64 format to use with flask.
    """
    fig = Figure(
        figsize=(
            float(plot_options["width"]) / DPI,
            float(plot_options["height"]) / DPI,
        ),
        dpi=DPI,
    )
    axis = fig.subplots()

    if len(solvers) < 2:
        output = BytesIO()
        fig.savefig(output, format="png")
        return base64.b64encode(output.getbuffer()).decode("ascii")

    profile_data = ProfileData(*solvers)
    for i, solver in enumerate(solvers):
        axis.plot(
            profile_data.breakpoints,
            profile_data.cumulative[:, i],
            label=solver.algname,
        )

    if plot_options["semilogx"]:
        axis.set_xscale("log")
    axis.set_xlim(profile_data.breakpoints.min(), profile_data.breakpoints.max())
    axis.set_ylim(-0.02, 1.04)
    axis.set_title(plot_options["title"])
    axis.set_xlabel(plot_options["xlabel"])
    axis.set_ylabel(plot_options["ylabel"])
    axis.legend()

    output = BytesIO()
    fig.savefig(output, format="png")
    return base64.b64encode(output.getbuffer()).decode("ascii")


perprof_web/app.py
"""Main file."""


import os

from flask import Flask, redirect, render_template, request, url_for
from perprof.solver_data import SolverData, read_table
from plotting import matplotlib_performance_profile
from werkzeug.utils import secure_filename

# Define folder to save uploaded files to process further
UPLOAD_FOLDER = os.path.join("uploads")

# Define allowed files (for this example I want only csv file)
ALLOWED_EXTENSIONS = {"csv", "table"}

app = Flask(__name__)
# Configure upload file path flask
app.config["UPLOAD_FOLDER"] = os.path.join(os.path.dirname(__file__), UPLOAD_FOLDER)

# Define secret key to enable session
app.secret_key = "This is your secret key to utilize session in Flask"

solvers = []
# For debugging purposes:
# solvers = [
#     read_table(os.path.join(app.config["UPLOAD_FOLDER"], "alpha.table")),
#     read_table(os.path.join(app.config["UPLOAD_FOLDER"], "beta.table")),
#     read_table(os.path.join(app.config["UPLOAD_FOLDER"], "gamma.table")),
# ]

plot_options = {
    "title": "Performance Profile",
    "xlabel": "Performance ratio",
    "ylabel": "Percentage of problems solved",
    "semilogx": "checked",
    "width": "800",
    "height": "500",
}


@app.route("/")
def home():
    """Home."""
    return render_template(
        "index.html",
        solvers=solvers,
    )


@app.route("/solvers", methods=("POST", "GET"))
def show_solvers():
    """Main page for solvers."""
    if request.method == "POST":
        action = request.form.get("action")

        if action == "upload":
            # upload file flask
            uploaded_file = request.files["uploaded-file"]
            filename = secure_filename(uploaded_file.filename)
            savepath = os.path.join(app.config["UPLOAD_FOLDER"], filename)
            uploaded_file.save(savepath)

            ext = filename.split(".")[-1]
            if ext == "csv":
                solvers.append(
                    SolverData(
                        f"Unnamed solver {len(solvers) + 1}",
                        savepath,
                    )
                )
            elif ext == "table":
                solvers.append(read_table(savepath))
            else:
                # TODO: Treat correctly using Flask
                raise ValueError(f"Bad {filename}. Unexpected extension {ext}.")
        elif action == "update":
            for field in ["title", "xlabel", "ylabel", "width", "height"]:
                current_value = plot_options[field]
                new_value = request.form.get(field)
                if new_value != current_value:
                    plot_options[field] = new_value

            plot_options["semilogx"] = request.form.get("semilogx")
        else:
            raise ValueError(f"Unexpected action {action}")

    plot_data = matplotlib_performance_profile(solvers, plot_options)

    return render_template(
        "solvers.html",
        plot_data=plot_data,
        plot_options=plot_options,
        solvers=solvers,
    )


@app.route("/solvers/<int:solver_id>")
def show_solver(solver_id):
    """Details of a solver."""
    return render_template(
        "solver.html",
        solvers=solvers,
        solver=solvers[solver_id],
    )


@app.route("/update/solver/<int:solver_id>", methods=["POST"])
def update_solver(solver_id):
    """Delete or update a solver."""
    action = request.form.get("action")

    if action == "delete":
        solvers.pop(solver_id)
    elif action == "edit":
        new_algname = request.form.get("algname")
        if len(new_algname.strip()) == 0:
            return "Solver name must be non-null"
        if new_algname != solvers[solver_id].algname:
            solvers[solver_id].algname = new_algname

    return redirect(url_for("show_solvers"))
